### PR TITLE
IR: Add DeadGPRStoreElimination

### PR DESCRIPTION
**Pull request by skmp**
This is based on the Flag Elimination pass, and is not optimal - but does eliminate quite a few dead stores.

### CODE REVIEW

**Comment by skmp** [comment]
*updated to use TrueBlock/FalseBlock instead of Arg[XX]*

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+          auto Op = IROp->CW<IR::IROp_StoreContext>();
+          if (IsFullGPR(Op->Offset, IROp->Size))
+            GPRMap[BlockNode].writes |= GPRBit(Op->Offset);
+          else
+            GPRMap[BlockNode].reads |= GPRBit(Op->Offset);
+        }
+        else if (IROp->Op == OP_STORECONTEXTINDEXED ||
+               IROp->Op == OP_LOADCONTEXTINDEXED ||
+               IROp->Op == OP_LOADCONTEXTPAIR ||
+               IROp->Op == OP_STORECONTEXTPAIR) {
```
Comment (Sonicadvance1) [inline External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp:79]: We also can't track through syscall here.
Syscall can modify full context state...
We should also probably move rax setting from syscall to be handled by syscall handler so we are more generic...

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+      }
+    }
+  }
+
+  // Pass 2
+  // Compute GPRs that are stored, but always ovewritten in the next blocks
+  // Propagate the information a few times to eliminate more
+  for (int i = 0; i < 5; i++)
```
Comment (Sonicadvance1) [inline External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp:95]: Make this constant a constexpr value towards the top for easy tinkering

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  {
+    for (auto [BlockNode, BlockIROp] : CurrentIR.GetBlocks()) {
+      for (auto [CodeNode, IROp] : CurrentIR.GetCode(BlockNode)) {
+
+        if (IROp->Op == OP_STORECONTEXT) {
+          auto Op = IROp->CW<IR::IROp_StoreContext>();
+          // If this OP_STORECONTEXT is never read, remove it
+          if (GPRMap[BlockNode].kill & GPRBit(Op->Offset)) {
+            IREmit->Remove(CodeNode);
+            //printf("Removed dead store %d, %d\n", Op->Offset, IROp->Size);
```
Comment (Sonicadvance1) [inline External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp:138]: remove debug printf?

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp (hunk 1 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp:79]: RCLSE, DeadFlags and RA assume that SYSCALLS don't modify the context. Do syscalls really do it?

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp (hunk 2 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp]: Done

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp (hunk 3 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp]: Done

**Review by Sonicadvance1 — APPROVED** [review]

File: External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp (hunk 1 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/IR/Passes/DeadGPRStoreElimination.cpp:82]: Will resolve later

**Pull request merged**